### ISSUE TITLE
Terminal takes a long time to start if `screen` is installed

### ISSUE THREAD

**Issue by bpmct** [coder/coder#11872]
https://github.com/coder/coder/assets/22407953/dabcb2d4-1849-4149-b060-c0fba2d67084

**Comment by matifali** [comment]
This is know and expected behavior. Of course not desirable. @mafredri and @code-asher have more context if we can remove/reduce this.

**Comment by mafredri** [comment]
For the best user experience, I would like for us to work on improving the Go rpty implementation. The problem with it is currently that it has no understanding of terminal modes, so if the terminal enters alternate mode and the instruction is evicted from the buffer, all instructions following that will be nonsensical.

@code-asher implemented the `screen` backend to work around that issue, and let `screen` do the "heavy lifting" (i.e. understanding terminal instructions).

But ultimately, if we have to stick with the separate backend, we could use another program as well, like `tmux`. I believe that also would lead to a better user experience.

But my main gripe with rpty is how behavior will change depending on programs you have installed. You might want screen installed on the system, but not the poor(er) performance from using it as a rpty backend. This is why I advocate for improving our Go implementation, it has the potential to "just work":tm:.

**Comment by code-asher** [comment]
tl;dr: we could re-implement xterm.js in Go?  Might be pretty easy actually, although could take some time.

----

Disclaimer: the terminal here is actually already started and ready to be used, screen just hides the prompt while showing that startup message in the lower left corner for some reason.  Typing will abort the timer and immediately show the prompt.  I think we can bypass this automatically but need to experiment to be sure.

But yeah, this is poor UX, and having to install screen is also a poor experience.

I did try tmux but I went with screen because when you scroll up in tmux it enters copy mode and you have to press `q` or scroll back to the bottom to be able to type again which I thought would be too confusing.  I think there may have been some other oddities around scrolling too.  But it might be worth revisiting and maybe there are some workarounds.

I think implementing a terminal emulator or multiplexer could be difficult and time-consuming, and ours may be very well worse than screen's, but it would be very, very nice to handle it all in Go.

The way VS Code does reconnections is to have a separate headless xterm.js instance running on the backend.  They feed it the same input, then when the client reconnects they ask the headless instance for its state and they send that to the client so it can restore said state.  Assuming we are still against bundling Node (which is fair), one path forward then would be to reimplement xterm.js in Go.  If we are just re-implementing 1:1 maybe it would actually be reasonably easy??

There is also https://pkg.go.dev/github.com/hinshun/vt10x but it was not clear to me how exactly to hook this up, and the implementation might be different from xterm.js anyway.

Otherwise I suppose we have to go with screen/tmux or roll our own.

**Comment by code-asher** [comment]
And just to be thorough: something else I tried was to not replay history at all, but when the client reconnects force the terminal to resize hoping it would re-draw the screen.  You would not have scrollback but at least you could see the current screen contents.

Unfortunately that did not work, but maybe there is a way to make that happen that I missed.

**Comment by code-asher** [comment]
One other path forward I would like to at least raise: we could remove reconnections and make users do it themselves with tmux/screen.  I think this is not what we want, but it does align with command-line ssh usage.
